iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/transformation/dag/transformer/multi/MappableUDFQueryRowTransformer.java

```diff
@@ -19,105 +19,50 @@
+import org.apache.tsfile.block.column.Column;
+import org.apache.tsfile.block.column.ColumnBuilder;
 
 public class MappableUDFQueryRowTransformer extends UDFQueryTransformer {
 
-  protected final LayerRowReader layerRowReader;
-  private final boolean isLayerRowReaderConstant;
+  protected final LayerReader layerReader;
 
-  public MappableUDFQueryRowTransformer(LayerRowReader layerRowReader, UDTFExecutor executor) {
+  public MappableUDFQueryRowTransformer(LayerReader layerReader, UDTFExecutor executor) {
     super(executor);
-    this.layerRowReader = layerRowReader;
-    this.isLayerRowReaderConstant = false;
-  }
-
-  @Override
-  protected boolean cacheValue() throws QueryProcessException, IOException {
-    if (!layerRowReader.next()) {
-      return false;
-    }
-
-    if (!isLayerRowReaderConstant) {
-      cachedTime = layerRowReader.currentTime();
-    }
-
-    if (layerRowReader.isCurrentNull()) {
-      currentNull = true;
-    } else {
-      Row row = layerRowReader.currentRow();
-      execute(row);
-    }
-
-    layerRowReader.readyForNext();
-    return true;
+    this.layerReader = layerReader;
   }
 
   @Override
   protected YieldableState yieldValue() throws Exception {
-    final YieldableState yieldableState = layerRowReader.yield();
+    final YieldableState yieldableState = layerReader.yield();
     if (!YieldableState.YIELDABLE.equals(yieldableState)) {
       return yieldableState;
     }
 
-    if (!isLayerRowReaderConstant) {
-      cachedTime = layerRowReader.currentTime();
-    }
-
-    if (layerRowReader.isCurrentNull()) {
-      currentNull = true;
-    } else {
-      Row row = layerRowReader.currentRow();
-      execute(row);
-    }
+    Column[] columns = layerReader.current();
+    cachedColumns = execute(columns);
 
-    layerRowReader.readyForNext();
+    layerReader.consumedAll();
     return YieldableState.YIELDABLE;
   }
 
-  private void execute(Row row) {
-    executor.execute(row);
-    Object currentValue = executor.getCurrentValue();
-    if (Objects.isNull(currentValue)) {
-      currentNull = true;
-      return;
-    }
-    switch (tsDataType) {
-      case INT32:
-        cachedInt = (int) currentValue;
-        break;
-      case INT64:
-        cachedLong = (long) currentValue;
-        break;
-      case FLOAT:
-        cachedFloat = (float) currentValue;
-        break;
-      case DOUBLE:
-        cachedDouble = (double) currentValue;
-        break;
-      case BOOLEAN:
-        cachedBoolean = (boolean) currentValue;
-        break;
-      case TEXT:
-        cachedBinary = (Binary) currentValue;
-        break;
-      default:
-        throw new UnSupportedDataTypeException(tsDataType.toString());
-    }
+  private Column[] execute(Column[] columns) {
+    int count = columns[0].getPositionCount();
+    ColumnBuilder valueColumnBuilder = TypeUtils.initColumnBuilder(tsDataType, count);
+
+    executor.execute(columns, valueColumnBuilder);
+
+    Column timeColumn = columns[1];
```

Comment:
Make sure that columns returned by layerReader only contains two columns.

Reply:
Done.